File: _515_find_largest_value_in_each_tree_row/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        result = []
        queue = deque([root])

        while queue:
            maximum = -math.inf
            length = len(queue)

            for _ in range(length):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)
                
                maximum = max(maximum, node.val)

            result.append(maximum)

        return result

File: _515_find_largest_value_in_each_tree_row/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        if (root == NULL) {
            return vector<int>();
        }

        vector<int> result;
        deque<TreeNode*> queue;
        queue.push_back(root);

        while (!queue.empty()) {
            int maximum = INT_MIN;
            int length = queue.size();

            for (int i = 0; i < length; i++) {
                TreeNode* node = queue.front();
                queue.pop_front();
                if (node->left != NULL) {
                    queue.push_back(node->left);
                }

                if (node->right != NULL) {
                    queue.push_back(node->right);
                }

                maximum = max(node->val, maximum);
            }

            result.push_back(maximum);
        }

        return result;
    }
};